docs/modules/ROOT/pages/references/architecture/metering-data-flow.adoc

```diff
@@ -0,0 +1,527 @@
+= Data model and data flow for Billing
+
+[abstract]
+====
+xref:appuio-cloud:ROOT:references/architecture/metering.adoc[Metering of resource usage] explains how the data collection for billing relevant data works.
+This page explains how that data gets transformed so that it can be ingested into a billing/ERP system.
+====
+
+The core of this part of the system has three responsibilites.
+First it must fetch the required data from the input system.
+Second that data must be enriched.
+That enrichment involves looking up prices and discounts from price tables.
+It also includes matching the identifiers from the source (Prometheus) with the identifiers from the target system.
+Third it must ingest that data into the target system.
+
+Key features of this component are:
+
+* It uses the Prometheus Query API as its source.
+* It allows to integrate any billing/ERP system as its target.
+  In order to achieve this, it must be loosely coupled.
+* Execution is idempotent.
+  If run twice for a given time frame with the same configuration, the result is the same.
+  However, if run again with a different configuration, the results reflect the changes made.
+  This ensures that errors in the past can easily be fixed.
+* Time aware configuration
+  Every configuration with the power to change the output, is made time aware.
+  This means that this config has timestamps attached denoting from when and until when that config is valid.
+
+== System idea
+
+This is an https://en.wikipedia.org/wiki/Extract,_transform,_load[ETL process].
+The extract and load part are well isolated and have only a dependency to their source/target system.
+The transform part on the other hand needs to bring parts of both the source and the target togehter.
+Instead of having one single ETL process, a multi stage model with an intermediate persistence will be used.
+The first stage fetches the data from the source into that intermediate storage.
+The second stage enriches that data by bringing in missing pieces from the target system.
+The third stage then ingests the data into the target system.
+This is done to keep this loosely coupled and thus simplify writing and maintaining the code base.
+It also allows to easily replace the code that talks to the target system.
+We take into account ease of replacing for code that talks to the target system to accommodate replacing the billing/ERP system and reselling of {product}.
```

Comment:
I agree that the data extraction and the ingestion into the external system should be as decoupled as possible. I'm not sure they are in this proposal. (If I understood it correctly) All three stages communicate directly through a shared database and the database contains target specific data.

What if for example the target systems id is not actually an int but a string? Or it's in another way not directly map-able?

I would suggest the ingestion is done by a completely different system and talks to the data extraction system through a clearly defined API (Databases are IMO a terrible API)

Reply:
Yes, during discussions I've also made the suggestion to go via API and not shared database. Atm, we are considering starting with shared database to get experience with the data model before we start designing an API that would constantly change while we collect experience.

Reply:
I don't see the advantage of a "constantly changing" shared database model compared to a "constantly changing" API.